### PR TITLE
Auto updates

### PR DESCRIPTION
**Pull request by jesseduffield**
The plan here is to have a user-configurable option deciding how frequently they want to check for updates and whether they want the latest build or the latest stable build. The user will also be able to configure whether these updates happen in the background or require a confirmation prompt.

By default the app will check for updates once every 7 days, with a prompt asking if you want to update.

Before deciding to check, a lastchecked timestamp is compared to the current time and the configured update frequency will be compared.

If a check occurs, the timestamp will be set to the current time

if the binary was not downloaded via a direct binary release from github, it will not update automatically.

I think I'm gonna have a thing for setting application status messages. There will be flash types which will just disappear after two seconds, and loading types which will disappear when a call is made to finish them. Loading types will have a loader animation on them, flash types won't. There will be a ranking of which things get shown i.e. updating comes first, then committing etc. 

To test locally you'll need to comment out the following guards:
```
if u.Config.GetVersion() == "unversioned" {
if u.majorVersionDiffers(u.Config.GetVersion(), newVersion)
if !userRequested && u.skipUpdateCheck() {
```

- [x] - fork go-getter to not raise an error on xml parse failure
- [x] - support updating the latest binary
- [x] - only update when checkForUpdates ldflag is used
- [x] - show updating status when update is in progress
- [x] - setup default config options ( frequency in days, never/background/prompt)
- [x] - ability to update from the status panel
- [x] - setup frontend prompt for update
- [x] - return focus to previous view after closing the auto update popup panel
- [x] - support auto-updating in the background

### CODE REVIEW

File: pkg/gui/gui.go [hunk 1]
```diff
@@ -318,6 +321,9 @@ func (gui *Gui) Run() error {
 	}
 	defer g.Close()
 
+	// TODO: do this more elegantly
+	gui.Updater.CheckForNewUpdate()
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.Updater.CheckForNewUpdate` is not checked

File: pkg/updates/updates.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	newVersion, err := u.getLatestVersionNumber()
+	if err != nil {
+		return "", err
+	}
+	u.NewVersion = newVersion
+	u.Log.Info("Current version is " + u.Config.GetVersion())
+	u.Log.Info("New version is " + newVersion)
+	if newVersion != u.Config.GetVersion() {
+		u.getBinaryUrl()
```
Comment (golangci) [inline pkg/updates/updates.go]: Error return value of `u.getBinaryUrl` is not checked

File: pkg/updates/updates.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		u.mappedArch(runtime.GOARCH),
+		extension,
+	)
+	u.Log.Info("url for latest release is " + url)
+	return url, nil
+}
+
+// worry about what happens if we fail halfway through or if the user closes the app halfway through
+
+func (u *Updater) downloadLatestBinary() error {
```
Comment (golangci) [inline pkg/updates/updates.go]: func `(*Updater).downloadLatestBinary` is unused

File: pkg/updates/updates.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	// attempt to download the binary with the url
+	// store in our config folder or other appropriate folder
+	return nil
+}
+
+func (u *Updater) installLatestBinary() error {
```
Comment (golangci) [inline pkg/updates/updates.go]: func `(*Updater).installLatestBinary` is unused

**Comment by jesseduffield** [comment]
yeah a flag called 'checkForUpdates' will do. I'll update the PR accordingly

File: pkg/gui/gui.go [hunk 5]
```diff
@@ -261,6 +264,19 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+		newVersion, err := gui.Updater.CheckForNewUpdate()
+		if err != nil {
+			return err
+		}
+		gui.Updater.NewVersion = "v0.1.75"
+		newVersion = "v0.1.75"
+		if newVersion != "" {
+			if err := gui.Updater.Update(); err != nil {
+				panic(err)
+				return err
```
Comment (golangci) [inline pkg/gui/gui.go]: unreachable code

File: pkg/gui/gui.go [hunk 6]
```diff
@@ -261,6 +264,19 @@ func (gui *Gui) layout(g *gocui.Gui) error {
 			return err
 		}
 
+		newVersion, err := gui.Updater.CheckForNewUpdate()
```
Comment (golangci) [inline pkg/gui/gui.go]: ineffectual assignment to `newVersion`

File: pkg/gui/gui.go [hunk 7]
```diff
@@ -277,18 +311,54 @@ func (gui *Gui) layout(g *gocui.Gui) error {
 	return nil
 }
 
+func (gui *Gui) onUpdateFinish(err error) error {
+	gui.State.Updating = false
+	gui.setAppStatus("")
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.setAppStatus` is not checked

File: pkg/gui/gui.go [hunk 8]
```diff
@@ -277,18 +311,54 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+	if err != nil {
+		gui.createErrorPanel(gui.g, "Update failed: "+err.Error())
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.createErrorPanel` is not checked

File: pkg/gui/gui.go [hunk 9]
```diff
@@ -277,18 +311,54 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+	}
+	if newVersion == "" {
+		return nil
+	}
+	title := "New version available!"
+	message := "Download latest version? (enter/esc)"
+	// TODO: support nil view in createConfirmationPanel or always go back to filesPanel when there is no previous view
+	return gui.createConfirmationPanel(gui.g, nil, title, message, func(g *gocui.Gui, v *gocui.View) error {
+		gui.State.Updating = true
+		gui.setAppStatus("updating...")
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.setAppStatus` is not checked

File: pkg/gui/gui.go [hunk 10]
```diff
@@ -277,18 +311,54 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+	viewNames := []string{"confirmation", "appStatus"}
+	for _, viewName := range viewNames {
+		if view, _ := g.View(viewName); view != nil {
+			content := gui.trimmedContent(view)
+			if strings.Contains(content, "...") {
+				staticContent := strings.Split(content, "...")[0] + "..."
+				if viewName == "appStatus" {
+					// panic(staticContent + " " + gui.loader())
+				}
+				gui.renderString(g, viewName, staticContent+" "+gui.loader())
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.renderString` is not checked

File: pkg/gui/gui.go [hunk 11]
```diff
@@ -277,18 +311,54 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
-		if strings.Contains(content, "...") {
-			staticContent := strings.Split(content, "...")[0] + "..."
-			gui.renderString(g, "confirmation", staticContent+" "+gui.loader())
+	viewNames := []string{"confirmation", "appStatus"}
+	for _, viewName := range viewNames {
+		if view, _ := g.View(viewName); view != nil {
+			content := gui.trimmedContent(view)
+			if strings.Contains(content, "...") {
+				staticContent := strings.Split(content, "...")[0] + "..."
+				if viewName == "appStatus" {
```
Comment (golangci) [inline pkg/gui/gui.go]: empty branch

File: pkg/updates/updates.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+		LastChecked: "today",
+		Log:         contextLogger,
+		Config:      config,
+		OSCommand:   osCommand,
+	}
+	return updater, nil
+}
+
+func (u *Updater) getLatestVersionNumber() (string, error) {
+	time.Sleep(5)
```
Comment (golangci) [inline pkg/updates/updates.go]: sleeping for 5 nanoseconds is probably a bug. Be explicit if it isn't: time.Sleep(5 * time.Nanosecond)

File: pkg/gui/gui.go [hunk 13]
```diff
@@ -137,6 +141,13 @@ func max(a, b int) int {
 	return b
 }
 
+func (gui *Gui) setAppStatus(status string) error {
```
Comment (golangci) [inline pkg/gui/gui.go]: func `(*Gui).setAppStatus` is unused

File: pkg/gui/gui.go [hunk 14]
```diff
@@ -277,18 +311,54 @@ func (gui *Gui) layout(g *gocui.Gui) error {
 	return nil
 }
 
+func (gui *Gui) onUpdateFinish(err error) error {
```
Comment (golangci) [inline pkg/gui/gui.go]: func `(*Gui).onUpdateFinish` is unused

File: pkg/gui/gui.go [hunk 15]
```diff
@@ -277,18 +311,54 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+	gui.State.Updating = false
+	gui.setAppStatus("")
+	if err != nil {
+		gui.createErrorPanel(gui.g, "Update failed: "+err.Error())
+	}
+	// TODO: on attempted quit, if downloading is true, ask if sure the user wants to quit
+	return nil
+}
+
+func (gui *Gui) onUpdateCheckFinish(newVersion string, err error) error {
```
Comment (golangci) [inline pkg/gui/gui.go]: func `(*Gui).onUpdateCheckFinish` is unused

File: pkg/config/app_config.go [hunk 16]
```diff
@@ -156,7 +224,23 @@ func getDefaultConfig() []byte {
...
+}
 
+// AppState stores data between runs of the app like when the last update check
+// was performed and which other repos have been checked out
+type AppState struct {
+	LastUpdateCheck int64
+	RecentRepos     []string
+}
+
+func getDefaultAppState() []byte {
```
Comment (golangci) [inline pkg/config/app_config.go]: `getDefaultAppState` is unused

File: pkg/gui/confirmation_panel.go [hunk 17]
```diff
@@ -93,17 +104,11 @@ func (gui *Gui) createConfirmationPanel(g *gocui.Gui, currentView *gocui.View, t
...
-			}
-			confirmationView.Title = title
-			confirmationView.FgColor = gocui.ColorWhite
-			gui.renderString(g, "confirmation", prompt)
-			gui.switchFocus(g, currentView, confirmationView)
-			return gui.setKeyBindings(g, handleConfirm, handleClose)
+		if err := gui.prepareConfirmationPanel(currentView, title, prompt); err != nil {
+			return err
 		}
+		gui.renderString(g, "confirmation", prompt)
```
Comment (golangci) [inline pkg/gui/confirmation_panel.go]: Error return value of `gui.renderString` is not checked

File: pkg/gui/gui.go [hunk 18]
```diff
@@ -250,6 +304,22 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
 	}
+
+	if appStatusView, err := g.SetView("appStatus", -1, optionsTop, width, optionsTop+2, 0); err != nil {
+		if err != gocui.ErrUnknownView {
+			return err
+		}
+		appStatusView.BgColor = gocui.ColorDefault
+		appStatusView.FgColor = gocui.ColorCyan
+		appStatusView.Frame = false
+		g.SetViewOnBottom("appStatus")
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `g.SetViewOnBottom` is not checked

File: pkg/gui/gui.go [hunk 19]
```diff
@@ -250,6 +304,22 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+	}
+
+	if appStatus != "" {
+		// might add this logic to the manager or somewhere better, just on a repeater
+		gui.renderString(gui.g, "appStatus", appStatus)
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.renderString` is not checked

File: pkg/gui/gui.go [hunk 20]
```diff
@@ -277,18 +349,53 @@ func (gui *Gui) layout(g *gocui.Gui) error {
 	return nil
 }
 
+func (gui *Gui) onUpdateFinish(err error) error {
+	gui.State.Updating = false
+	gui.statusManager.removeStatus("updating")
+	if err != nil {
+		gui.createErrorPanel(gui.g, "Update failed: "+err.Error())
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.createErrorPanel` is not checked

File: pkg/gui/gui.go [hunk 21]
```diff
@@ -277,18 +349,53 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+	viewNames := []string{"confirmation", "appStatus"}
+	for _, viewName := range viewNames {
+		if view, _ := g.View(viewName); view != nil {
+			content := gui.trimmedContent(view)
+			if strings.Contains(content, "...") {
+				staticContent := strings.Split(content, "...")[0] + "..."
+				if viewName == "appStatus" {
+					// panic(staticContent + " " + utils.Loader())
+				}
+				gui.renderString(g, viewName, staticContent+" "+utils.Loader())
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.renderString` is not checked

File: pkg/gui/status_panel.go [hunk 22]
```diff
@@ -45,9 +45,16 @@ func (gui *Gui) renderStatusOptions(g *gocui.Gui) error {
 	return gui.renderOptionsMap(g, map[string]string{
 		"o": gui.Tr.SLocalize("OpenConfig"),
 		"e": gui.Tr.SLocalize("EditConfig"),
+		"u": gui.Tr.SLocalize("CheckForUpdate"),
 	})
 }
 
+func (gui *Gui) handleCheckForUpdate(g *gocui.Gui, v *gocui.View) error {
+	gui.Updater.CheckForNewUpdate(gui.onUpdateCheckFinish, true)
+	gui.createMessagePanel(gui.g, nil, "Checking for updates", "Checking for updates...")
```
Comment (golangci) [inline pkg/gui/status_panel.go]: Error return value of `gui.createMessagePanel` is not checked

File: pkg/config/app_config.go [hunk 23]
```diff
@@ -72,73 +93,120 @@ func (c *AppConfig) GetName() string {
...
+	folder := configDirs.QueryFolderContainsFile(filename)
 	if folder == nil {
-		// create the file as an empty config and load it
+		// create the file as empty
 		folders := configDirs.QueryFolders(configdir.Global)
-		if err := folders[0].WriteFile("config.yml", []byte{}); err != nil {
-			return err
+		if err := folders[0].WriteFile(filename, []byte{}); err != nil {
+			panic(err.Error())
+			return "", err
```
Comment (golangci) [inline pkg/config/app_config.go]: unreachable code

File: pkg/gui/gui.go [hunk 24]
```diff
@@ -277,23 +344,60 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
 
 func (gui *Gui) updateLoader(g *gocui.Gui) error {
-	if confirmationView, _ := g.View("confirmation"); confirmationView != nil {
-		content := gui.trimmedContent(confirmationView)
+	if view, _ := g.View("confirmation"); view != nil {
+		content := gui.trimmedContent(view)
 		if strings.Contains(content, "...") {
 			staticContent := strings.Split(content, "...")[0] + "..."
-			gui.renderString(g, "confirmation", staticContent+" "+gui.loader())
+			gui.renderString(g, "confirmation", staticContent+" "+utils.Loader())
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.renderString` is not checked

File: pkg/gui/status_panel.go [hunk 25]
```diff
@@ -45,9 +45,16 @@ func (gui *Gui) renderStatusOptions(g *gocui.Gui) error {
 	return gui.renderOptionsMap(g, map[string]string{
 		"o": gui.Tr.SLocalize("OpenConfig"),
 		"e": gui.Tr.SLocalize("EditConfig"),
+		"u": gui.Tr.SLocalize("CheckForUpdate"),
 	})
 }
 
+func (gui *Gui) handleCheckForUpdate(g *gocui.Gui, v *gocui.View) error {
+	gui.Updater.CheckForNewUpdate(gui.onUpdateCheckFinish, true)
+	gui.createMessagePanel(gui.g, v, "Checking for updates", "Checking for updates...")
```
Comment (golangci) [inline pkg/gui/status_panel.go]: Error return value of `gui.createMessagePanel` is not checked

File: pkg/gui/gui.go [hunk 26]
```diff
@@ -277,23 +303,71 @@ func (gui *Gui) layout(g *gocui.Gui) error {
 	return nil
 }
 
+func (gui *Gui) onUpdateFinish(err error) error {
+	gui.State.Updating = false
+	gui.statusManager.removeStatus("updating")
+	gui.renderString(gui.g, "appStatus", "")
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.renderString` is not checked

File: pkg/gui/gui.go [hunk 27]
```diff
@@ -277,23 +303,71 @@ func (gui *Gui) layout(g *gocui.Gui) error {
...
+	if err != nil {
+		gui.createErrorPanel(gui.g, "Update failed: "+err.Error())
```
Comment (golangci) [inline pkg/gui/gui.go]: Error return value of `gui.createErrorPanel` is not checked

**Comment by jesseduffield** [comment]
@dawidd6 I'm making it that the binary needs a particular LDFLAG variable set at build time to make it do automatic update checks. I'm guessing your packages are building themselves rather than pointing at the github release files? Because if that's the case then I think we're good

**Comment by dawidd6** [comment]
So, can i just omit this variable and updates will be disabled?

**Comment by jesseduffield** [comment]
yep :)

**Pull request merged**